From 3f97bfb7d382dc1bd21375b2c1ac6d5dd5a4aab9 Mon Sep 17 00:00:00 2001
From: Brendan King <Brendan.King@imgtec.com>
Date: Fri, 27 Jul 2018 15:16:58 +0100
Subject: [PATCH 08/16] modesetting: add option to not render a SW cursor
Upstream-Status: Pending

If option NoCursor is set to true in the Device section of xorg.conf,
don't render a software cursor.

This is intended for benchmarking on a system with no hardware cursor
support. Rendering a software cursor can have a significant impact
on performance.

The name of the option was chosen to match that of a similar option
in the IMG PVR video driver.
---
 hw/xfree86/drivers/modesetting/driver.c          | 6 ++++++
 hw/xfree86/drivers/modesetting/driver.h          | 1 +
 hw/xfree86/drivers/modesetting/drmmode_display.c | 3 +++
 hw/xfree86/drivers/modesetting/drmmode_display.h | 1 +
 4 files changed, 11 insertions(+)

diff --git a/hw/xfree86/drivers/modesetting/driver.c b/hw/xfree86/drivers/modesetting/driver.c
index 535f49d1d..6056251a9 100644
--- a/hw/xfree86/drivers/modesetting/driver.c
+++ b/hw/xfree86/drivers/modesetting/driver.c
@@ -145,6 +145,7 @@ static const OptionInfoRec Options[] = {
     {OPTION_VARIABLE_REFRESH, "VariableRefresh", OPTV_BOOLEAN, {0}, FALSE},
     {OPTION_USE_GAMMA_LUT, "UseGammaLUT", OPTV_BOOLEAN, {0}, FALSE},
     {OPTION_ASYNC_FLIP_SECONDARIES, "AsyncFlipSecondaries", OPTV_BOOLEAN, {0}, FALSE},
+    {OPTION_NOSW_CURSOR, "NoCursor", OPTV_BOOLEAN, {0}, FALSE},
     {-1, NULL, OPTV_NONE, {0}, FALSE}
 };
 
@@ -1158,6 +1159,11 @@ PreInit(ScrnInfoPtr pScrn, int flags)
         ms->drmmode.sw_cursor = TRUE;
     }
 
+    if (xf86ReturnOptValBool(ms->drmmode.Options, OPTION_NOSW_CURSOR, FALSE)) {
+        ms->drmmode.sw_cursor = FALSE;
+        ms->drmmode.nosw_cursor = TRUE;
+    }
+
     ms->cursor_width = 64;
     ms->cursor_height = 64;
     ret = drmGetCap(ms->fd, DRM_CAP_CURSOR_WIDTH, &value);
diff --git a/hw/xfree86/drivers/modesetting/driver.h b/hw/xfree86/drivers/modesetting/driver.h
index 71aa8730e..ec07cf5f1 100644
--- a/hw/xfree86/drivers/modesetting/driver.h
+++ b/hw/xfree86/drivers/modesetting/driver.h
@@ -61,6 +61,7 @@ typedef enum {
     OPTION_VARIABLE_REFRESH,
     OPTION_USE_GAMMA_LUT,
     OPTION_ASYNC_FLIP_SECONDARIES,
+    OPTION_NOSW_CURSOR,
 } modesettingOpts;
 
 typedef struct
diff --git a/hw/xfree86/drivers/modesetting/drmmode_display.c b/hw/xfree86/drivers/modesetting/drmmode_display.c
index 48dccad73..681176c42 100644
--- a/hw/xfree86/drivers/modesetting/drmmode_display.c
+++ b/hw/xfree86/drivers/modesetting/drmmode_display.c
@@ -1677,6 +1677,9 @@ drmmode_set_cursor(xf86CrtcPtr crtc)
         xf86CrtcConfigPtr xf86_config = XF86_CRTC_CONFIG_PTR(crtc->scrn);
         xf86CursorInfoPtr cursor_info = xf86_config->cursor_info;
 
+        if (drmmode_crtc->drmmode->nosw_cursor)
+                return TRUE;
+
         cursor_info->MaxWidth = cursor_info->MaxHeight = 0;
         drmmode_crtc->drmmode->sw_cursor = TRUE;
     }
diff --git a/hw/xfree86/drivers/modesetting/drmmode_display.h b/hw/xfree86/drivers/modesetting/drmmode_display.h
index 29f9b8f7d..20130e64f 100644
--- a/hw/xfree86/drivers/modesetting/drmmode_display.h
+++ b/hw/xfree86/drivers/modesetting/drmmode_display.h
@@ -102,6 +102,7 @@ typedef struct {
     drmEventContext event_context;
     drmmode_bo front_bo;
     Bool sw_cursor;
+    Bool nosw_cursor;
 
     /* Broken-out options. */
     OptionInfoPtr Options;
-- 
2.34.1

